### ISSUE TITLE
FR: ESP32 library

### ISSUE THREAD

**Issue by mailinger-mate** [tailscale/tailscale#5220]
### What are you trying to do?

Hello, do you think Tailscale could build and publish a library for the Espressif ESP32 microcontroller that is widely used in IoT applications?

### How should we solve this?

There is a library that mplements WireGuard on ESP32 already and it could be a good starting point:
https://github.com/ciniml/WireGuard-ESP32-Arduino

### What is the impact of not solving this?

The current workaround is to use a Tailscale compatible router and create a subnet for the WiFi enabled MCUs.

**Comment by DentonGentry** [comment]
We publish a guide to get tailscale binaries down to <5 Megs. This isn't small enough for ESP32, and still contains a bunch of Linux dependencies, but is a start. We think we can get it a Meg smaller by omitting more stuff.
https://tailscale.com/kb/1207/small-tailscale/

Generating ESP32 binaries from Go code looks like it would require tinygo, which we haven't used but could be tried.

**Comment by seguidor777** [comment]
I haven't tried it but here is [another implementation](https://github.com/trombik/esp_wireguard) for ESP-IDF where esp32 is supported

**Comment by orent** [comment]
RP2040 has a dual core ARM cpu. Some boards have 8 or 16 MB flash and wifi - all for a single digit $price tag.

I think it would make a great reference platform for deeply embedded Tailscale.

**Comment by yousaf465** [comment]
https://github.com/esphome/feature-requests/issues/1444 some work already done

**Comment by DentonGentry** [comment]
Even once a wireguard implementation is available, the difficulty will be in Tailscale's Discovery protocol for NAT traversal.

**Comment by yousaf465** [comment]
> Even once a wireguard implementation is available, the difficulty will be in Tailscale's Discovery protocol for NAT traversal.

So there is no remote possibility of it working on a small SOC with esps3 already available in 16mb version.

**Comment by DentonGentry** [comment]
It seems possible, https://tailscale.com/kb/1207/small-tailscale/ can compile the whole thing down to < 5 Megs including both wireguard-go and the Discovery protocol.

To set expectations though: Tailscale Inc is unlikely to undertake this work for ESP32.

**Comment by yousaf465** [comment]
> It seems possible, https://tailscale.com/kb/1207/small-tailscale/ can compile the whole thing down to < 5 Megs including both wireguard-go and the Discovery protocol.
> 
> To set expectations though: Tailscale Inc is unlikely to undertake this work for ESP32.
obviously, the open-source community will do the work, but any guidance from the tail-scale devs will help a lot.

**Comment by rjjrbatarao** [comment]
I'am also interested on this, I saw recently that tailscale have c library? https://tailscale.dev/blog/libtailscale

**Comment by MrBuddyCasino** [comment]
This is just a C API for their golang code, so that won’t fly.

**Comment by MrBuddyCasino** [comment]
> I'am also interested on this, I saw recently that tailscale have c library?

This is just a C API for their golang code, so that won’t fly.

**Comment by zekageri** [comment]
oh man it would be good

**Comment by dragetd** [comment]
Is anyone working on this?

There are a few steps that need to be tackled. Luckily, some of the harder issues already have the right puzzle pieces. And with the modern versions of the ESP which have 8 or even 16mb of flash, there is plenty of room.

As mentioned above, there are a few implementation for the ESP32 based on the lwIP Implementation for Wireguard. It (Wireguard) looks surprisingly doable when I checked the code.

To get tailscale running, simply a small binary is not enough (especially upx is not helpful here)

I could imagine taking the following approach:
- Strip tailscale of all the functions not feasible on an uC. Webinterface, maybe even the whole oauth2 flow, benchmark features, port-forwarding, tunneling etc.
- Port remaining code to tinyGo. In the first step, this would most likely mean commenting out / stubbing 90% of the remaining tailscale functionality down to the core logic to get something (useless) compiled that runs on the ESP32 platform. It would be still a huge step.
- Then slowly work on each bit of commented out code and try to find alternative ways to do it on the ESP. Hooking into the existing wireguard implementation, managing IP/firewall etc. will all need completely new solutions, as the ESP is not a Linux system.

I have no tinyGo experience, just looked into the project. Also did not yet check out the tailscale codebase. But these things look less firefighting that porting wireguard or the compile target for go… and since those steps are already done, it might be possible.

This is still a lot of work. But I'd be up for a shared hackathon on this, seeing how far we can get. :D

**Comment by ozanerturk** [comment]
Here is a repo that implements wireguard for esp-idf. https://github.com/trombik/esp_wireguard
Knowing Tailscale is based on wireguard, my assumption is it would be  a good starting point.
What do you say ?

**Comment by dopeytree** [comment]
Would be good to have it working on a smaller device for sure. How practical it is on esp32 is TBC.

It may be that it always needs a node (node has 4g sim card for example) but perhaps this could be factored in somehow so that the esp32 only has to do core tailscale and the node controller can do the actual internet communication. 

Sort of a bit like the way 'matter' works over ip6 locally but the tailscale way would mean devices could talk to each other other the internet securely.

**Comment by drio** [comment]
> We publish a guide to get tailscale binaries down to <5 Megs.

Would it be possible to subscribe to doc updates or new docs via rss the same way we do for the blog? There is ton of great material in the kb that I don't want to miss.

**Comment by drio** [comment]
I know this thread is about getting TS running specifically in the esp32 but I'd like to point out that you can run TS in [gokrazy](https://github.com/gokrazy/). You don't need a "full OS" and it runs very well on the [Pi Zero 2W](https://www.raspberrypi.com/products/raspberry-pi-zero-2-w/).

**Comment by kdpuvvadi** [comment]
@drio  Pi  Zero 2 is for more powerfull than any or all of the esps

**Comment by rsemihkoca** [comment]
Gonna leave a comment since it is considerably hot topic for me. This update would change iot world

**Comment by tckb** [comment]
super interested in this. would be great way to make my groundstation remote

**Comment by StephensGraeme** [comment]
This would be a bit of a game changer for many!

**Comment by rjjrbatarao** [comment]
i dont know what nat traversal used on go source code but here is one using webtrc working on esp32 https://github.com/sepfy/libpeer

**Comment by bradfitz** [comment]
As one workaround, see https://github.com/tailscale/tailscale/tree/lp/cmd/lopower#readme

**Comment by dragetd** [comment]
One major blocker (apart from all the work itself) is: https://github.com/tinygo-org/tinygo/issues/2024

No wifi support for tinygo. =/

**Comment by CoolhappyDE** [comment]
I have to leave a comment as well, it would be so nice to have Tailscale support the ESP platform 😍

**Comment by HWXLR8** [comment]
@ Avery Pennarun I will literally quit my job and work full-time on making this real just pay for my rent and 1 pack of ramen noodles per day (will share with cat). PLEASE REPLY

**Comment by tckb** [comment]
IRL, I was abe to use a variant of ESP Wireguard to connect to a wireguard network on ESP32. If anyone is just interested on a private network this should be enough.  if you want to have a tailscale net, you can either setup a peer that is on a tailscale net and foward the pakets - this needs you to be a network ninja to understand and implement all the ip*-foo that is needed. 

PS: I am a network ninja

**Comment by hitecSmartHome** [comment]
Any update on this?

**Comment by alfs** [comment]
I've started a tailscale port from golang to c/c++ as a proof of concept. Runs on the ESP32-C3 SuperMini, probably other ESP32 as well. Requires about 1,3 MB Flash and fits in 320 kB RAM.

It connects to headscale control plane, gets peers and keys, passing all the http2, ts2021 and noice protocols. Work in progress but I think it's come quite far.

Code at https://github.com/alfs/tailscale-iot

**Comment by StraightfaceStudios** [comment]
The new ability to setup customer tailnets within an organization would make iot much more interesting. 

Largest deployment issue would seem to be how chatty Tailscale is consuming lots of bandwidth per month which wouldn’t be a good fit for something with a cellular modem.

**Comment by beastbroak30** [comment]
I am working on that and I guess, the binaries can be compiled  in less, but errors struck, but anyways

**Comment by Csontikka** [comment]
For anyone still following this — we built an ESPHome component that puts native Tailscale on ESP32 using the microlink library:
  https://github.com/Csontikka/esphome-tailscale

**Comment by yousaf465** [comment]
https://discord.com/channels/429907082951524364/1493243779379892234

Discord link

**Comment by Csontikka** [comment]
Sibling project - turns an ESP32-S3 into a full WiFi NAT + Tailscale subnet routerconfigured from a web UI.
https://github.com/Csontikka/esp32-tailscale-subnet-router
